adam_core/utils/indexable.py

```diff
@@ -1,174 +1,451 @@
-            return len(self.index.unique(level="class_index"))
-        else:
-            err = "Length is not defined for this class."
-            raise NotImplementedError(err)
+    Returns
+    -------
+    values_mapped : np.ndarray[int]
+        The values mapped to integers.
+    """
+    # We use pandas since numpy unique is slower for object dtypes
+    df = pd.DataFrame({"values": values})
+    df_unique = df.drop_duplicates(keep="first")
+    df_unique["values_mapped"] = np.arange(0, len(df_unique))
+    values_mapped = df.merge(df_unique, on="values", how="left")
+
+    return np.array(values_mapped["values_mapped"])
+
+
+class Indexable:
+    """
+    Class that enables indexing and slicing of itself and its members.
+    If an Indexable subclass has members that are `~numpy.ndarray`s, `~numpy.ma.core.MaskedArray`s,
+    lists, or `~astropy.time.core.Time`s then these members are appropriately sliced and
+    indexed along their first axis. Any members that are dicts, OrderedDicts, floats,
+    integers or strings are not indexed and left unchanged.
+
+    Indexable maintains two array indexes:
+        - class: the externally facing index (ie. the index into the class itself)
+        - members: the index of the members of the class (ie. the index into
+            the data structures carried by the class)
+
+
+    Class Index:
+    [0, 1, 2]
+    The class index is used to determine the length (size) of the class. In this example,
+    the len(instance) would be 3.
+
+    Class To Members Mapping:
+    [0, 0, 0, 1, 1, 1, 2, 2, 2]
+    If, as above, the class to members mapping is sorted (monotonically increasing), then
+    it is converted to an array of slices which allows for faster querying.
+    [slice(0, 2), slice(2, 5), slice(5, 8)]
+
+    Members Index:
+    [0, 1, 2, 3, 4, 5, 6, 7, 8]
+
+    For example, here orbits is a subclass of Indexable and has members that are
+    `~numpy.ndarray`s, `~numpy.ma.core.MaskedArray`s, and `~astropy.time.core.Time`s.
+
+    ```
+    from adam_core.orbits import Orbits
+    from adam_core.backend import PYOORB
+    from astropy.time import Time
+    from astropy import units as u
+
+    # Instantiate an Orbits object with 3 orbits
+    t0 = Time([59000.0], scale="utc", format="mjd")
+    orbits = Orbits.from_horizons(["Eros", "Ceres", "Duende"], t0)
+    len(orbits) == 3
+
+    # If I only wanted one orbit
+    single_orbit = orbits[0] # Orbits with one orbit
+
+    # Lets remove the other ones
+    del orbits[1:]
+
+    # Lets propagate the orbits to 10 new times
+    orbits = Orbits.from_horizons(["Eros", "Ceres", "Duende"], t0)
+    t1 = t0 + np.arange(0, 10) * u.day
+    backend = PYOORB()
+    propagated_orbits = backend.propagate_orbits(orbits, t1)
+
+    # The underlying arrays are now 10 * 3 in length
+    len(propagated_orbits) == 30
+
+    # If we update the index to be on orbit_id
+    # then the length of the class is 3 even though each
+    # member is 10 * 3 in length
```

Comment:
My head is really spinning on this... the idea is that you get a deduplicated index to the unique values? Yeah, I think I see that as I read the rest of the code more carefully. It's definitely not how I normally think of data, which is basically as tables. This is some other data structure, I'm not sure what I'd call it. I guess a hashmap of arrays, sort of, or something.

Feel free to just resolve this musing - I'm only posting it to record my thought process because this is pretty mind-bending and probably should get more explanation in the docs.